R/step_hcai_missing.R

```diff
@@ -0,0 +1,144 @@
+#'                                          size = n, replace = TRUE),
+#'             disease = ifelse(hemoglobin_count < 15, "Yes", "No")
+#' )
+#' 
+#' # Initialize
+#' my_recipe <- recipe(disease ~ ., data = d)
+#' 
+#' # Create recipe
+#' my_recipe <- my_recipe %>%
+#'   step_hcai_missing(all_nominal())
+#' my_recipe
+#' 
+#' # Train recipe
+#' trained_recipe <- prep(my_recipe, training = d)
+#' 
+#' # Apply recipe
+#' data_modified <- bake(trained_recipe, newdata = d)
+#' 
+step_hcai_missing <- function(recipe,
+                              ...,
+                              role = NA,
+                              trained = FALSE,
+                              na_percentage = NULL) {
+  terms <- quos(...)
+  if (length(terms) == 0)
+    stop("Please supply at least one variable specification. See ?selections.")
+  add_step(
+    recipe,
+    step_hcai_missing_new(
+      terms = terms,
+      role = role,
+      trained = trained,
+      na_percentage = na_percentage
+    )
+  )
+}
+
+# Initialze a new object
+step_hcai_missing_new <- function(terms = NULL,
+                                  role = NA,
+                                  trained = FALSE,
+                                  na_percentage = NULL) {
+  step(
+    subclass = "hcai_missing",
+    terms = terms,
+    role = role,
+    trained = trained,
+    na_percentage = na_percentage
+  )
+}
+
+#' @export
+prep.step_hcai_missing <- function(x, training, info = NULL, ...) {
+  col_names <- terms_select(terms = x$terms, info = info)
+  na_percentage <- sapply(training[, col_names], function(x) {
+    100 * round(sum(is.na(x)) / length(x), 3)
+  }
+  )
+
+  # Give warnings about greater than 50% null
+  if (any(na_percentage > 50)) {
+  message("The following categorical columns have greater than 50% missing values 
+and will be filled with the category 'hcai_missing':")
+  message(paste0(names(na_percentage[na_percentage > 50]),
+                 ": ", 
+                 na_percentage[na_percentage > 50], "%",
+                 collapse = "\n"))
+  }
+  
+  step_hcai_missing_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    na_percentage = na_percentage
+  )
+}
+
+#' @importFrom tidyr replace_na
+#' @importFrom stats setNames
+#' @export
+bake.step_hcai_missing <- function(object, newdata, ...) {
+  vars <- names(object$na_percentage)
+
+  # Add new level to all factors
+  newdata[vars] <- lapply(newdata[vars], function(x){
+    levels(x) <- c(levels(x), "hcai_missing")
+    x
+  })
+
+  # Replace NAs
+  replacement_list <-
+    rep("hcai_missing", length(vars)) %>%
+    as.list %>%
+    setNames(vars)
+  newdata <- newdata %>%
+    replace_na(replacement_list)
+}
+
+#' @importFrom utils getFromNamespace
+#' @export
+print.step_hcai_missing <-
+  function(x, width = max(20, options()$width - 30), ...) {
+    printer <- getFromNamespace("printer", "recipes")
+    cat("Filling NA with hcai_missing for ", sep = "")
```

Comment:
Would it be hard to print the fraction of entries that are being filled in for each column?

Reply:
That's what the percentage filled is. I'm going to add a `tidy` method that shows that nicely.